# 2107541 Machine Learning for Environmental Applications
## Lecture 6 : Data Preprocessing & Supervised Learning - Regression II & Model Evaluation
### Done By : 6730084521 Chatrphol Ovanonchai

## Step 1: Import the Dataset

In [1]:
# import libraries
import pandas as pd
import numpy as np 
import matplotlib.pyplot as pyplot
import seaborn as sns

In [2]:
df = pd.read_csv("air_quality_regression.csv")
df.head(5)

,PM2.5 (ug/m3),PM10 (ug/m3),NO2 (ug/m3),CO (mg/m3),O3 (ug/m3),Temperature (C),Humidity (%),Wind Speed (m/s),AQI
0,59.3,83.9,31.5,0.91,64.6,17.1,61.9,4.2,131
1,142.9,207.6,46.7,1.69,69.8,12.7,71.7,4.9,290
2,111.1,165.6,39.0,1.38,55.5,27.6,42.7,5.6,221
3,91.8,127.4,25.7,1.14,94.8,20.8,36.6,1.8,186
4,27.6,71.7,15.6,0.50,85.0,32.7,51.9,2.7,57


## Step 2 : Explore the Dataset

In [3]:
# Check dataset size
df.shape

(220, 9)

In [4]:
# Check data types and missing values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 220 entries, 0 to 219
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   PM2.5 (ug/m3)     220 non-null    float64
 1   PM10 (ug/m3)      220 non-null    float64
 2   NO2 (ug/m3)       220 non-null    float64
 3   CO (mg/m3)        220 non-null    float64
 4   O3 (ug/m3)        220 non-null    float64
 5   Temperature (C)   220 non-null    float64
 6   Humidity (%)      220 non-null    float64
 7   Wind Speed (m/s)  220 non-null    float64
 8   AQI               220 non-null    int64  
dtypes: float64(8), int64(1)
memory usage: 15.6 KB


In [5]:
# Check missing values
df.isnull().sum()

PM2.5 (ug/m3)       0
PM10 (ug/m3)        0
NO2 (ug/m3)         0
CO (mg/m3)          0
O3 (ug/m3)          0
Temperature (C)     0
Humidity (%)        0
Wind Speed (m/s)    0
AQI                 0
dtype: int64

In [6]:
# Descriptive statistics
df.describe()

,PM2.5 (ug/m3),PM10 (ug/m3),NO2 (ug/m3),CO (mg/m3),O3 (ug/m3),Temperature (C),Humidity (%),Wind Speed (m/s),AQI
count,220.000000,220.000000,220.000000,220.000000,220.000000,220.000000,220.000000,220.000000,220.000000
mean,75.128182,114.367727,29.515455,1.000318,73.976818,23.871364,62.794091,3.247727,150.745455
std,43.096784,58.845559,15.415261,0.419000,12.799139,7.204105,15.492199,1.628610,83.869320
min,5.800000,10.000000,3.000000,0.280000,44.800000,12.100000,35.300000,0.600000,20.000000
25%,37.950000,62.325000,17.075000,0.660000,64.900000,17.100000,49.100000,1.875000,78.750000
50%,76.700000,108.750000,27.800000,1.000000,73.000000,23.900000,64.700000,3.250000,145.500000
75%,114.775000,165.450000,41.925000,1.327500,81.750000,29.925000,74.500000,4.700000,224.000000
max,148.100000,227.000000,67.200000,1.780000,109.200000,35.900000,90.000000,6.000000,303.000000


## Step 3: Define X (Feature) and Y (Target)

In [7]:
# features
features = df.columns[:-1] # AQI not included

X = df[features] # feature
y = df['AQI'] # target

In [8]:
features

Index(['PM2.5 (ug/m3)', 'PM10 (ug/m3)', 'NO2 (ug/m3)', 'CO (mg/m3)',
       'O3 (ug/m3)', 'Temperature (C)', 'Humidity (%)', 'Wind Speed (m/s)'],
      dtype='object')

## Step 4: Train/Test Split 

In [9]:
from sklearn.model_selection import train_test_split

X_train , X_test , y_train , y_test = train_test_split(X , y , test_size = 0.2 , random_state = 42)

print("Training Shape:" , X_train.shape)
print("Testing Shape:" , X_test.shape)

Training Shape: (176, 8)
Testing Shape: (44, 8)


## Step 5.1 : Multiple Linear Regression (Model Training)

In [10]:
from sklearn.linear_model import LinearRegression
linear_model = LinearRegression()

linear_model.fit(X = X_train , y = y_train)

LinearRegression()

In [11]:
y_pred_linear = linear_model.predict(X_test) 
y_pred_linear

array([ 44.16051946,  30.70751987, 173.63495421, 112.30822432,
        67.15770357, 269.10778313, 152.24998506, 153.85045798,
       221.11515679, 165.3969826 ,  21.39298855, 172.83597912,
        42.71089926, 230.35247063, 243.92253077, 129.65738207,
        69.66497222, 136.60927143, 255.42021599,  62.36310546,
        68.9098926 , 160.25015442, 169.69395938, 241.77050435,
       270.43814672,   4.73080414, 132.79173639,  93.6118115 ,
       205.15797479, 281.98984712,  58.3816604 , 266.67972016,
        87.82357713, 213.8419233 , 249.83479532,  31.48238586,
       112.53443183, 116.68763743,  87.60985945, 191.45698277,
       200.71524138, 184.54929969, 233.39946808, 146.92735393])

## Step 5.2 : Multiple Linear Regression (Evaluation)

In [12]:
## check coefficients
linear_model.coef_

array([ 7.60320737e-01,  9.50108165e-02,  6.93556058e-01,  8.46810363e+01,
        6.15703193e-02,  1.38464239e-02,  1.06658766e-01, -1.86092463e-01])

In [13]:
## evaluation
from sklearn.metrics import (r2_score,mean_absolute_error,mean_squared_error)

r2 = r2_score(y_test, y_pred_linear)
mae = mean_absolute_error(y_test, y_pred_linear)
mse = mean_squared_error(y_test, y_pred_linear)
rmse = np.sqrt(mse)

print(f"R² = {r2:.3f}")
print(f"MAE = {mae:.3f}")
print(f"MSE = {mse:.3f}")
print(f"RMSE = {rmse:.3f}")

R² = 0.939
MAE = 13.916
MSE = 400.170
RMSE = 20.004


## Step 6.1 : Polynomial Regression (Model Training)

In [14]:
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline

## create pipeline
poly_model_pipeline = Pipeline( [ 
                      ("polynomial" , PolynomialFeatures(degree = 2 , include_bias = False , interaction_only = False)) , 
                            # build polynomial feature first -> (name , model)
                      ("regression" , LinearRegression())
                       ] )

## train model pipeline woth fit 
poly_model_pipeline.fit(X_train , y_train)

Pipeline(steps=[('polynomial', PolynomialFeatures(include_bias=False)),
                ('regression', LinearRegression())])

In [15]:
## predict value 
y_predicted_poly = poly_model_pipeline.predict(X_test)

/opt/anaconda3/lib/python3.13/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: divide by zero encountered in matmul
  return X @ coef_ + self.intercept_
/opt/anaconda3/lib/python3.13/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: overflow encountered in matmul
  return X @ coef_ + self.intercept_
/opt/anaconda3/lib/python3.13/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: invalid value encountered in matmul
  return X @ coef_ + self.intercept_


## Step 6.2 : Polynomial Regression (Model Evaluation)

In [16]:
from sklearn.metrics import (mean_absolute_error , mean_squared_error , r2_score)

r2_poly = r2_score(y_true = y_test , y_pred = y_predicted_poly)
mae_poly = mean_absolute_error(y_true = y_test , y_pred = y_predicted_poly)
mse_poly = mean_squared_error(y_true = y_test , y_pred = y_predicted_poly)
rmse_poly = np.sqrt(mse_poly)

print(f"R² = {r2_poly:.3f}")
print(f"MAE = {mae_poly:.3f}")
print(f"MSE = {mse_poly:.3f}")
print(f"RMSE = {rmse_poly:.3f}")

R² = 0.802
MAE = 20.521
MSE = 1301.883
RMSE = 36.082


## Step 6.3 : Polynomial Regression (Find Each Degree Evaluation)

In [17]:
degrees = [1, 2, 3, 4, 5]
for degree in degrees:
    model = Pipeline([
    ("polynomial", PolynomialFeatures(degree=degree,include_bias=False)),
    ("regression", LinearRegression())
    ])
    
    model.fit(X_train, y_train)
 
    train_pred = model.predict(X_train)
    test_pred = model.predict(X_test)
    train_r2 = r2_score(y_train, train_pred)
    test_r2 = r2_score(y_test, test_pred)
    
    print(
            f"Degree {degree}: "
            f"Train R² = {train_r2:.3f}, "
            f"Test R² = {test_r2:.3f}"
    )

Degree 1: Train R² = 0.931, Test R² = 0.939
Degree 2: Train R² = 0.972, Test R² = 0.802
Degree 3: Train R² = 0.999, Test R² = -62.153
Degree 4: Train R² = 1.000, Test R² = -56.066
Degree 5: Train R² = 1.000, Test R² = -35.494


/opt/anaconda3/lib/python3.13/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: divide by zero encountered in matmul
  return X @ coef_ + self.intercept_
/opt/anaconda3/lib/python3.13/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: overflow encountered in matmul
  return X @ coef_ + self.intercept_
/opt/anaconda3/lib/python3.13/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: invalid value encountered in matmul
  return X @ coef_ + self.intercept_
/opt/anaconda3/lib/python3.13/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: divide by zero encountered in matmul
  return X @ coef_ + self.intercept_
/opt/anaconda3/lib/python3.13/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: overflow encountered in matmul
  return X @ coef_ + self.intercept_
/opt/anaconda3/lib/python3.13/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: invalid value encountered in matmul
  return X @ coef_ + self.intercept_


## Step 7 : Ridge Regression

In [18]:
## disable RuntimeWarning
import warnings

# Suppress only RuntimeWarnings
warnings.filterwarnings("ignore", category=RuntimeWarning)

In [19]:
# import sklearn
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Ridge
from sklearn.model_selection import GridSearchCV

# build ridge model pipeline
ridge_model_pipeline = Pipeline([
                            ("scaler" , StandardScaler()) , # do scaling first 
                            ("regression" , Ridge()) # then do ridge regression (linear regression with penalty L2)
] , memory = None)

# create parameter grid 
ridge_param_grid = { 
                    "regression__alpha" : [0.001 , 0.01 , 0.1 , 1 , 10 , 100] # get parameter by usign _+_
                    }

## grid search cv 
ridge_grid_cv_model = GridSearchCV(estimator = ridge_model_pipeline , param_grid = ridge_param_grid , 
                          cv = 5 , scoring = "r2" # using R^2 to score , with 5-fold CV 
)

## train model cv
ridge_grid_cv_model.fit(X = X_train , y = y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('regression', Ridge())]),
             param_grid={'regression__alpha': [0.001, 0.01, 0.1, 1, 10, 100]},
             scoring='r2')

In [20]:
print("Best Ridge alpha:", ridge_grid_cv_model.best_params_)

Best Ridge alpha: {'regression__alpha': 10}


In [21]:
## predict target
y_pred_ridge = ridge_grid_cv_model.predict(X_test)

In [22]:
## model evaluation
r2_ridge = r2_score(y_test, y_pred_ridge)
mae_ridge = mean_absolute_error(y_test, y_pred_ridge)
mse_ridge = mean_squared_error(y_test, y_pred_ridge)
rmse_ridge = np.sqrt(mse_ridge)

print(f"R² = {r2_ridge:.3f}")
print(f"MAE = {mae_ridge:.3f}")
print(f"MSE = {mse_ridge:.3f}")
print(f"RMSE = {rmse_ridge:.3f}")

R² = 0.938
MAE = 13.858
MSE = 405.858
RMSE = 20.146


## Step 8 : Lasso Regression

In [23]:
## import lasso 
from sklearn.linear_model import Lasso

## create pipeline
lasso_model_pipeline = Pipeline([
                            ("scaler" , StandardScaler()) , # do scaling first 
                            ("regression" , Lasso(max_iter = 10000)) # then do lasso regression (linear regression with penalty L1)
                            ## max_iter = so much to prevent a convergence warning
] , memory = None)

## create parameter grid
lasso_param_grid = {"regression__alpha": [0.001, 0.01, 0.1, 1, 10, 100]}

## do grid search cv
lasso_grid_cv_model = GridSearchCV( estimator = lasso_model_pipeline, param_grid = lasso_param_grid,
                          cv=5,scoring="r2")

In [24]:
## train cv with .fit()
lasso_grid_cv_model.fit(X = X_train , y = y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('regression', Lasso(max_iter=10000))]),
             param_grid={'regression__alpha': [0.001, 0.01, 0.1, 1, 10, 100]},
             scoring='r2')

In [25]:
print("Best Lasso alpha:", lasso_grid_cv_model.best_params_)

Best Lasso alpha: {'regression__alpha': 1}


In [26]:
## predict data 
y_pred_lasso = lasso_grid_cv_model.predict(X_test)

In [27]:
## model evaluation
r2_lasso = r2_score(y_test, y_pred_lasso)
mae_lasso = mean_absolute_error(y_test, y_pred_lasso)
mse_lasso = mean_squared_error(y_test, y_pred_lasso)
rmse_lasso = np.sqrt(mse_lasso)

print(f"R² = {r2_lasso:.3f}")
print(f"MAE = {mae_lasso:.3f}")
print(f"MSE = {mse_lasso:.3f}")
print(f"RMSE = {rmse_lasso:.3f}")

R² = 0.941
MAE = 13.695
MSE = 390.250
RMSE = 19.755


## Step 9 : Elastic Net

In [28]:
from sklearn.linear_model import ElasticNet

elastic_model_pipeline = Pipeline([
                                ("scaler", StandardScaler()),
                                ("regression", ElasticNet(max_iter=10000))
                            ] , memory = None)

elastic_param_grid = {
 "regression__alpha": [0.001, 0.01, 0.1, 1, 10], ## L2
 "regression__l1_ratio": [0.1, 0.3, 0.5, 0.7, 0.9] ## L1
}

elastic_grid_cv_model = GridSearchCV(estimator = elastic_model_pipeline, param_grid = elastic_param_grid,
                            cv=5 , scoring="r2")

## train model
elastic_grid_cv_model.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('regression',
                                        ElasticNet(max_iter=10000))]),
             param_grid={'regression__alpha': [0.001, 0.01, 0.1, 1, 10],
                         'regression__l1_ratio': [0.1, 0.3, 0.5, 0.7, 0.9]},
             scoring='r2')

In [29]:
print("Best Elastic Net parameters:", elastic_grid_cv_model.best_params_)

Best Elastic Net parameters: {'regression__alpha': 1, 'regression__l1_ratio': 0.9}


In [30]:
## predict data
y_pred_elastic = elastic_grid_cv_model.predict(X_test)

In [31]:
## model evaluation
r2_elastic = r2_score(y_test, y_pred_elastic)
mae_elastic = mean_absolute_error(y_test, y_pred_elastic)
mse_elastic = mean_squared_error(y_test, y_pred_elastic)
rmse_elastic = np.sqrt(mse_elastic)

print(f"R² = {r2_elastic:.3f}")
print(f"MAE = {mae_elastic:.3f}")
print(f"MSE = {mse_elastic:.3f}")
print(f"RMSE = {rmse_elastic:.3f}")

R² = 0.935
MAE = 14.286
MSE = 424.443
RMSE = 20.602


## Step 10: Compare all models

In [32]:
results = pd.DataFrame({
 "Model": ["Linear Regression","Polynomial Regression","Ridge","Lasso","Elastic Net"] ,
 "R²": [r2, r2_poly, r2_ridge, r2_lasso, r2_elastic],
 "MAE": [mae, mae_poly, mae_ridge, mae_lasso, mae_elastic],
 "RMSE": [rmse, rmse_poly, rmse_ridge, rmse_lasso, rmse_elastic]
})
results.round(4).sort_values(by = ["R²","MAE","RMSE"] , ascending = [False , False , False])

,Model,R²,MAE,RMSE
3,Lasso,0.9405,13.6948,19.7547
0,Linear Regression,0.9390,13.9164,20.0043
2,Ridge,0.9381,13.8583,20.1459
4,Elastic Net,0.9353,14.2860,20.6020
1,Polynomial Regression,0.8015,20.5209,36.0816


## Extra: Compare Coefficients

In [33]:
# Get coefficients from each model
linear_coefficients = linear_model.coef_

ridge_coefficients = (
 ridge_grid_cv_model.best_estimator_
 .named_steps["regression"]
 .coef_ )

lasso_coefficients = (
 lasso_grid_cv_model.best_estimator_
 .named_steps["regression"]
 .coef_ )

elastic_coefficients = (
 elastic_grid_cv_model.best_estimator_
 .named_steps["regression"]
 .coef_ )

# Create comparison table
coef_table = pd.DataFrame({
 "Feature": features,
 "Linear Regression": linear_coefficients,
 "Ridge": ridge_coefficients,
 "Lasso": lasso_coefficients,
 "Elastic Net": elastic_coefficients
})
coef_table.round(3)

,Feature,Linear Regression,Ridge,Lasso,Elastic Net
0,PM2.5 (ug/m3),0.760,28.973,32.769,27.154
1,PM10 (ug/m3),0.095,13.242,5.053,15.070
2,NO2 (ug/m3),0.694,12.612,10.310,12.848
3,CO (mg/m3),84.681,28.684,35.329,26.653
4,O3 (ug/m3),0.062,-0.177,0.000,-0.102
5,Temperature (C),0.014,0.320,0.000,0.000
6,Humidity (%),0.107,1.689,0.773,1.000
7,Wind Speed (m/s),-0.186,-0.238,-0.000,-0.000


## -- Code below has been done by AI , so to do next is optimizing data --

## Step 11 : Data Cleaning
ทุกโมเดลใน Step 10 ได้ R² ไม่เกิน ~0.94 ไม่ว่าจะเปลี่ยนโมเดลแบบไหน → น่าจะมีปัญหาที่ **ข้อมูล** ไม่ใช่ที่โมเดล

ลองดูแถวที่โมเดลทายพลาดมากที่สุด (residual)

In [34]:
# ให้ Linear Regression ทายทุกแถว แล้วดู 10 แถวที่พลาดมากที่สุด
error_table = df.assign(Predicted = linear_model.predict(X).round(1))
error_table["Error"] = (y - error_table["Predicted"]).round(1)

error_table.loc[error_table["Error"].abs().sort_values(ascending=False).index].head(10)

,PM2.5 (ug/m3),PM10 (ug/m3),NO2 (ug/m3),CO (mg/m3),O3 (ug/m3),Temperature (C),Humidity (%),Wind Speed (m/s),AQI,Predicted,Error
208,8.0,91.0,57.0,1.00,71.0,27.0,66.0,5.0,264,116.8,147.2
203,84.0,103.0,17.0,1.00,102.0,34.0,66.0,3.0,249,150.4,98.6
207,28.0,131.0,52.0,1.00,71.0,31.0,84.0,2.0,38,134.9,-96.9
211,95.0,12.0,62.0,1.00,84.0,22.0,85.0,4.0,274,181.9,92.1
210,111.0,195.0,4.0,1.00,74.0,23.0,75.0,3.0,104,169.7,-65.7
205,38.0,44.0,20.0,1.00,45.0,28.0,36.0,2.0,40,105.3,-65.3
209,83.0,93.0,23.0,1.00,105.0,15.0,80.0,5.0,92,153.9,-61.9
219,129.0,100.0,49.0,1.00,99.0,32.0,79.0,5.0,152,207.3,-55.3
216,27.0,97.0,13.0,1.00,96.0,20.0,46.0,3.0,150,101.0,49.0
197,135.1,191.8,41.0,1.52,63.1,32.8,59.8,2.1,303,255.4,47.6


**สังเกต:** แถวที่ทายพลาดมากเกือบทั้งหมดอยู่ที่ index 202–219 ซึ่งเป็นข้อมูลเสีย เพราะ
- CO = 1.00 พอดีทุกแถว เท่ากับค่าเฉลี่ยของ CO → น่าจะเป็นค่าที่หายไปแล้วถูกเติมด้วยค่าเฉลี่ย
- ทุกค่าเป็นเลขกลม และบางแถว PM10 < PM2.5 (แถว 211: PM2.5 = 95, PM10 = 12) ซึ่งเป็นไปไม่ได้ทางฟิสิกส์
- AQI ไม่สัมพันธ์กับมลพิษ (แถว 208: PM2.5 = 8 แต่ AQI = 264)

→ ลบแถวเหล่านี้ทิ้ง แล้ว train ทุกโมเดลใหม่

In [35]:
from sklearn.base import clone

# ลบแถวเสีย (CO = 1.00 พอดี)
bad_rows = df["CO (mg/m3)"] == 1.0
df_clean = df[~bad_rows].reset_index(drop=True)
print("Before:", df.shape, "-> After:", df_clean.shape)

# โมเดลเดิมจาก Step 5-9
all_models = {
    "Linear Regression"    : linear_model,
    "Polynomial Regression": poly_model_pipeline,
    "Ridge"                : ridge_grid_cv_model,
    "Lasso"                : lasso_grid_cv_model,
    "Elastic Net"          : elastic_grid_cv_model,
}

def get_params(m):
    # alpha / l1-ratio ที่ GridSearchCV เลือก (Linear / Polynomial ไม่มี → NaN)
    p = getattr(m, "best_params_", {})
    return {"alpha": p.get("regression__alpha", np.nan), "l1-ratio": p.get("regression__l1_ratio", np.nan)}

def evaluate_all(cols):
    # split เหมือน Step 4 แล้ว train ทุกโมเดลใหม่บนข้อมูลที่ clean แล้ว (clone = copy setting แต่ยังไม่ train)
    X_tr, X_te, y_tr, y_te = train_test_split(df_clean[cols], df_clean["AQI"], test_size=0.2, random_state=42)
    rows = []
    for name, m in all_models.items():
        fitted = clone(m).fit(X_tr, y_tr)
        y_pred_c = fitted.predict(X_te)
        rows.append({"Model": name,
                     "R²"   : r2_score(y_te, y_pred_c),
                     "MAE"  : mean_absolute_error(y_te, y_pred_c),
                     "RMSE" : np.sqrt(mean_squared_error(y_te, y_pred_c)),
                     **get_params(fitted)})
    return pd.DataFrame(rows).set_index("Model")

# เทียบกับผลก่อน clean จาก Step 10
results_clean = evaluate_all(features)
before = results.set_index("Model")
pd.DataFrame({"R² (before)"  : before["R²"],   "R² (clean)"  : results_clean["R²"],
              "RMSE (before)": before["RMSE"], "RMSE (clean)": results_clean["RMSE"]}).round(4)

Before: (220, 9) -> After: (202, 9)


,R² (before),R² (clean),RMSE (before),RMSE (clean)
Model,,,,
Linear Regression,0.9390,0.9776,20.0043,12.1726
Polynomial Regression,0.8015,0.9680,36.0816,14.5456
Ridge,0.9381,0.9774,20.1459,12.2255
Lasso,0.9405,0.9772,19.7547,12.2861
Elastic Net,0.9353,0.9773,20.6020,12.2571


**ผล:** ลบแถวเสีย 18 แถว แล้ว**ทุกโมเดลดีขึ้นหมด** เช่น Linear Regression R² ขึ้นจาก 0.939 → 0.978 และ RMSE ลดจาก 20.0 → 12.2 ส่วน Polynomial ดีขึ้นมากที่สุดเพราะ degree 2 ไวต่อ outlier มาก

## Step 12 : Feature Selection (บนข้อมูลที่ clean แล้ว)
มี feature 8 ตัว อยากรู้ว่าตัวไหนช่วยทำนาย AQI จริง → เทียบหลายชุด feature ด้วย 5-fold Cross-Validation และให้ Forward Selection เลือกให้

In [36]:
from sklearn.model_selection import cross_validate, KFold
from sklearn.feature_selection import SequentialFeatureSelector

# แบ่งข้อมูล 5 ส่วน สลับกันเป็น test แล้วเฉลี่ยคะแนน
cv = KFold(n_splits=5, shuffle=True, random_state=42)

feature_sets = {
    "PM2.5 only"      : ["PM2.5 (ug/m3)"],
    "CO only"         : ["CO (mg/m3)"],
    "PM2.5 + NO2"     : ["PM2.5 (ug/m3)", "NO2 (ug/m3)"],
    "PM2.5 + CO + NO2": ["PM2.5 (ug/m3)", "CO (mg/m3)", "NO2 (ug/m3)"],
    "All 8 features"  : list(features),
}

cv_results = []
for name, cols in feature_sets.items():
    scores = cross_validate(LinearRegression(), df_clean[cols], df_clean["AQI"], cv=cv,
                            scoring=["r2", "neg_root_mean_squared_error"])
    cv_results.append({"Features": name,
                       "CV R²"   : scores["test_r2"].mean(),
                       "CV RMSE" : -scores["test_neg_root_mean_squared_error"].mean()})   # sklearn ให้ค่าติดลบ
display(pd.DataFrame(cv_results).round(4).sort_values("CV R²", ascending=False))

# Forward Selection: เริ่มจากชุดว่าง แล้วเพิ่ม feature ที่ช่วยเพิ่ม CV R² มากที่สุดทีละตัว
for k in [1, 2, 3]:
    sfs = SequentialFeatureSelector(LinearRegression(), n_features_to_select=k,
                                    direction="forward", cv=cv, scoring="r2")
    sfs.fit(df_clean[features], df_clean["AQI"])
    print(k, "features:", list(features[sfs.get_support()]))

,Features,CV R²,CV RMSE
2,PM2.5 + NO2,0.9808,11.6486
4,All 8 features,0.9808,11.6401
3,PM2.5 + CO + NO2,0.9804,11.7558
0,PM2.5 only,0.9785,12.3433
1,CO only,0.9621,16.3591


1 features: ['PM2.5 (ug/m3)']
2 features: ['PM2.5 (ug/m3)', 'NO2 (ug/m3)']
3 features: ['PM2.5 (ug/m3)', 'NO2 (ug/m3)', 'Wind Speed (m/s)']


**ผล:** PM2.5 สำคัญที่สุด (ตัวเดียวได้ CV R² = 0.979 แล้ว) และ **PM2.5 + NO2** ได้ CV R² = 0.981 เท่ากับใส่ครบ 8 ตัว แต่ใช้แค่ 2 ตัว Forward Selection (k = 2) ก็เลือกชุดเดียวกัน

> CO ดูสำคัญที่สุดถ้าใช้ข้อมูลเดิม แต่เป็นผลจากแถวเสีย (CO = 1.00 คงที่) พอ clean แล้ว CO only เหลือ 0.962

Train ทุกโมเดลใหม่ด้วย PM2.5 + NO2

In [37]:
# train ทุกโมเดลใหม่ด้วย feature ที่เลือก
selected_features = ["PM2.5 (ug/m3)", "NO2 (ug/m3)"]

results_selected = evaluate_all(selected_features)
results_selected.round(4).sort_values("R²", ascending=False)

,R²,MAE,RMSE,alpha,l1-ratio
Model,,,,,
Polynomial Regression,0.9782,10.1298,12.0087,NaN,NaN
Linear Regression,0.9762,10.7617,12.5546,NaN,NaN
Ridge,0.9762,10.7621,12.5549,0.001,NaN
Lasso,0.9762,10.7576,12.5439,0.100,NaN
Elastic Net,0.9761,10.7660,12.5587,0.001,0.9


### สรุป Step 11–12

In [38]:
# best model ของแต่ละ step พร้อม hyperparameter ที่ GridSearchCV เลือก
params = pd.DataFrame({name: get_params(m) for name, m in all_models.items()}).T
results_before = results.set_index("Model").join(params)

stages = {
    "ก่อน clean (Step 10)"                : ("8 ตัว", results_before),
    "หลัง clean (Step 11)"                : ("8 ตัว", results_clean),
    "หลัง clean + เลือก feature (Step 12)": ("PM2.5 + NO2", results_selected),
}

summary = pd.DataFrame([
    {"Stage": stage, "Features": feats, "Best Model": table["R²"].idxmax(),
     **table.loc[table["R²"].idxmax(), ["R²", "MAE", "RMSE", "alpha", "l1-ratio"]].to_dict()}
    for stage, (feats, table) in stages.items()
]).set_index("Stage")
summary.round(4)

,Features,Best Model,R²,MAE,RMSE,alpha,l1-ratio
Stage,,,,,,,
ก่อน clean (Step 10),8 ตัว,Lasso,0.9405,13.6948,19.7547,1.0,NaN
หลัง clean (Step 11),8 ตัว,Linear Regression,0.9776,10.5193,12.1726,NaN,NaN
หลัง clean + เลือก feature (Step 12),PM2.5 + NO2,Polynomial Regression,0.9782,10.1298,12.0087,NaN,NaN


- **ปัญหาอยู่ที่ข้อมูล ไม่ใช่โมเดล:** การ clean ข้อมูลทำให้ R² ขึ้นจาก ~0.94 → ~0.98 ส่วนการเปลี่ยนโมเดลแทบไม่ต่างกัน (ทุกโมเดล 0.976–0.978)
- **ใช้แค่ 2 features (PM2.5 + NO2) ได้ผลพอๆ กับ 8 ตัว** → ตัด PM10, CO, O3, Temperature, Humidity, Wind Speed ออกได้

**บทเรียน:** ก่อนเปลี่ยนโมเดลให้ซับซ้อนขึ้น ให้ดู residual ก่อนว่ามีแถวไหนผิดปกติไหม